File: android/core/src/test/kotlin/mu/core/OAuthFixtureTest.kt
package mu.core

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long
import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue
import kotlin.test.fail

/**
 * OAuth + PKCE 契約測試：跑 contract/fixtures/oauth_cases/（provider.md §10）。
 * 另含 RefreshingTokenSource 的行為測試（fake transport）。
 */
class OAuthFixtureTest {

    @Test
    fun `all oauth fixture cases match expected json byte for byte`() {
        val casesDir = findDir("contract/fixtures/oauth_cases") ?: fail("oauth_cases not found")
        val names = casesDir.listFiles()?.filter { it.isDirectory }?.map { it.name }?.sorted()
            ?: fail("cannot list $casesDir")
        assertTrue(names.size >= 6, "expected >=6 oauth cases, got ${names.size}")

        val failures = ArrayList<String>()
        for (name in names) {
            val caseDir = File(casesDir, name)
            val expected = File(caseDir, "expected.json").readBytes()
            val script = Json.parseToJsonElement(File(caseDir, "script.json").readText()).jsonObject
            val out = ArrayList<Map<String, Any?>>()
            for (el in script["entries"]!!.jsonArray) {
                val e = el.jsonObject
                fun str(k: String) = e[k]?.jsonPrimitive?.contentOrNull ?: ""
                val kind = str("type")
                val entryName = str("name")
                val row = LinkedHashMap<String, Any?>()
                when (kind) {
                    "authorize" -> {
                        val config = config(str("provider"), str("clientId"), str("redirectUri"))
                        row["challenge"] = OAuth.codeChallenge(str("verifier"))
                        row["kind"] = kind
                        row["name"] = entryName
                        row["url"] = OAuth.authorizationUrl(config, str("verifier"), str("state"))
                    }
                    "redirect" -> {
                        val r = OAuth.parseRedirect(str("url"), str("state"))
                        val result = LinkedHashMap<String, Any?>()
                        r.code?.let { result["code"] = it }
                        r.error?.let { result["error"] = it }
                        result["status"] = r.status.key
                        row["kind"] = kind
                        row["name"] = entryName
                        row["result"] = result
                    }
                    "form" -> {
                        val config = OAuth.Config("", "", "", emptyList(), str("clientId"), str("redirectUri"))
                        row["exchange"] = OAuth.exchangeForm(config, str("code"), str("verifier"))
                        row["kind"] = kind
                        row["name"] = entryName
                        row["refresh"] = OAuth.refreshForm(config, str("refreshToken"))
                    }
                    "token_response" -> {
                        val now = e["nowMs"]!!.jsonPrimitive.long
                        val state = TokenState.parse(e["prev"]?.jsonPrimitive?.contentOrNull)
                            .applying(str("body"), now)
                        row["kind"] = kind
                        row["name"] = entryName
                        row["needsRefresh"] = state.needsRefresh(now)
                        row["serialized"] = state.serialize()
                        row["state"] = stateMap(state)
                        row["usable"] = state.isUsable(now)
                    }
                    "expiry" -> {
                        val now = e["nowMs"]!!.jsonPrimitive.long
                        val state = TokenState.parse(str("state"))
                        row["kind"] = kind
                        row["name"] = entryName
                        row["needsRefresh"] = state.needsRefresh(now)
                        row["usable"] = state.isUsable(now)
                    }
                    "error" -> {
                        row["class"] = OAuth.classifyTokenError(e["status"]!!.jsonPrimitive.int, str("body")).key
                        row["kind"] = kind
                        row["name"] = entryName
                    }
                    else -> fail("unknown entry type [$kind]")
                }
                out.add(row)
            }
            val actual = CanonicalJson.render(out).toByteArray(Charsets.UTF_8)
            if (!expected.contentEquals(actual)) {
                failures.add("case [$name]\n--- expected ---\n${expected.decodeToString()}\n" +
                    "--- actual ---\n${actual.decodeToString()}")
            }
        }
        if (failures.isNotEmpty()) {
            fail("${failures.size}/${names.size} oauth cases drifted:\n\n" +
                failures.joinToString("\n\n========\n\n"))
        }
    }

    @Test
    fun `generated verifier is valid and distinct`() {
        val a = OAuth.makeCodeVerifier()
        assertEquals(64, a.length)
        assertNotEquals(a, OAuth.makeCodeVerifier())
        val allowed = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~".toSet()
        assertTrue(a.all { it in allowed })
        assertEquals(43, OAuth.makeCodeVerifier(10).length)
        assertEquals(128, OAuth.makeCodeVerifier(999).length)
        assertEquals(43, OAuth.makeState().length)
    }

    @Test
    fun `refreshing token source`() {
        val config = OAuth.Config.gdrive("cid", "music.mu.android:/oauth2redirect")
        val fake = FakeTokenServer()
        val persisted = ArrayList<TokenState>()
        val sleeps = ArrayList<Long>()
        var clock = 1_700_000_000_000L
        val source = RefreshingTokenSource(
            config, fake,
            TokenState("at-1", "rt-1", clock + 3_600_000),
            now = { clock }, sleep = { sleeps.add(it) }, persist = { persisted.add(it) },
        )

        assertEquals("at-1", source.token())
        assertEquals(0, fake.requests.size, "未過期不打網路")

        clock += 3_600_000
        fake.responses.add(200 to """{"access_token":"at-2","expires_in":3599}""")
        assertEquals("at-2", source.token())
        assertEquals("rt-1", source.current.refreshToken, "回應無 refresh_token → 沿用")
        assertEquals(1, persisted.size)
        assertEquals(
            "client_id=cid&grant_type=refresh_token&refresh_token=rt-1",
            fake.requests[0].body.toString(Charsets.UTF_8),
        )

        clock += 3_600_000
        fake.responses.addAll(listOf(
            503 to "", 429 to "", 200 to """{"access_token":"at-3","expires_in":3599}""",
        ))
        assertEquals("at-3", source.token())
        assertEquals(listOf(1000L, 2000L), sleeps, "§2.1 退避序列")

        clock += 3_600_000
        fake.responses.add(400 to """{"error":"invalid_grant"}""")
        assertFailsWith<ProviderException.Auth> { source.token() }

        val empty = RefreshingTokenSource(config, fake, TokenState.EMPTY, now = { clock }, sleep = {})
        assertFailsWith<ProviderException.Auth> { empty.token() }
    }

    @Test
    fun `exchange flow`() {
        val config = OAuth.Config.dropbox("cid", "http://127.0.0.1:7777/callback")
        val fake = FakeTokenServer()
        fake.responses.add(200 to """{"access_token":"at","refresh_token":"rt","expires_in":14400}""")
        val source = RefreshingTokenSource(config, fake, TokenState.EMPTY,
            now = { 1_700_000_000_000L }, sleep = {})
        val state = source.exchange("the-code", "the-verifier")
        assertEquals("at", state.accessToken)
        assertEquals("rt", state.refreshToken)
        assertEquals(1_700_000_000_000L + 14_400_000, state.expiresAtMs)
        assertEquals(
            "client_id=cid&code=the-code&code_verifier=the-verifier&grant_type=authorization_code" +
                "&redirect_uri=http%3A%2F%2F127.0.0.1%3A7777%2Fcallback",
            fake.requests[0].body.toString(Charsets.UTF_8),
        )
    }

    private fun config(provider: String, clientId: String, redirectUri: String) = when (provider) {
        "gdrive" -> OAuth.Config.gdrive(clientId, redirectUri)
        "dropbox" -> OAuth.Config.dropbox(clientId, redirectUri)
        else -> fail("unknown provider $provider")
    }

    private fun stateMap(s: TokenState) = linkedMapOf<String, Any?>(
        "accessToken" to s.accessToken,
        "expiresAtMs" to s.expiresAtMs,
        "refreshToken" to s.refreshToken,
        "scope" to s.scope,
    )

    private fun findDir(rel: String): File? {
        var dir: File? = File(System.getProperty("user.dir")).absoluteFile
        while (dir != null) {
            val c = File(dir, rel)
            if (c.isDirectory) return c
            dir = dir.parentFile
        }
        return null
    }
}

/** 腳本化 token 端點（空佇列 = 200 `{}`）。 */
class FakeTokenServer : HttpTransport {
    val responses = ArrayDeque<Pair<Int, String>>()
    val requests = ArrayList<HttpRequest>()

    override fun send(req: HttpRequest): HttpResponse {
        requests.add(req)
        val (status, body) = responses.removeFirstOrNull() ?: (200 to "{}")
        return HttpResponse(status, body.toByteArray(Charsets.UTF_8))
    }
}

File: apple/CrateCore/Tests/CrateCoreTests/OAuthFixtureTest.swift
import XCTest
@testable import CrateCore

/// OAuth + PKCE 契約測試：跑 contract/fixtures/oauth_cases/（provider.md §10）。
/// 另含 RefreshingTokenSource 的行為測試（fake transport）。
final class OAuthFixtureTest: XCTestCase {

    func testAllOAuthFixtureCasesMatchByteForByte() throws {
        let casesDir = try XCTUnwrap(findDir("contract/fixtures/oauth_cases"), "oauth_cases not found")
        let names = try FileManager.default.contentsOfDirectory(atPath: casesDir.path)
            .filter { name -> Bool in
                var isDir: ObjCBool = false
                let ok = FileManager.default.fileExists(
                    atPath: casesDir.appendingPathComponent(name).path, isDirectory: &isDir)
                return ok && isDir.boolValue
            }
            .sorted()
        XCTAssertGreaterThanOrEqual(names.count, 6)

        var failures: [String] = []
        for name in names {
            let caseDir = casesDir.appendingPathComponent(name)
            let expected = try String(contentsOf: caseDir.appendingPathComponent("expected.json"),
                                      encoding: .utf8)
            let script = try XCTUnwrap(
                JSONSerialization.jsonObject(
                    with: try Data(contentsOf: caseDir.appendingPathComponent("script.json")))
                as? [String: Any])
            var out: [CanonicalJson.JSONValue] = []
            for e in try XCTUnwrap(script["entries"] as? [[String: Any]]) {
                let kind = e["type"] as? String ?? ""
                let entryName = e["name"] as? String ?? ""
                var pairs: [(String, CanonicalJson.JSONValue)] = []
                switch kind {
                case "authorize":
                    let config = try makeConfig(provider: e["provider"] as? String ?? "",
                                                clientId: e["clientId"] as? String ?? "",
                                                redirectUri: e["redirectUri"] as? String ?? "")
                    let verifier = e["verifier"] as? String ?? ""
                    pairs = [
                        ("challenge", .string(OAuth.codeChallenge(verifier))),
                        ("kind", .string(kind)),
                        ("name", .string(entryName)),
                        ("url", .string(OAuth.authorizationUrl(
                            config: config, verifier: verifier,
                            state: e["state"] as? String ?? ""))),
                    ]
                case "redirect":
                    let r = OAuth.parseRedirect(url: e["url"] as? String ?? "",
                                                expectedState: e["state"] as? String ?? "")
                    var resultPairs: [(String, CanonicalJson.JSONValue)] = []
                    if let code = r.code { resultPairs.append(("code", .string(code))) }
                    if let err = r.error { resultPairs.append(("error", .string(err))) }
                    resultPairs.append(("status", .string(r.status.rawValue)))
                    pairs = [("kind", .string(kind)), ("name", .string(entryName)),
                             ("result", .object(resultPairs))]
                case "form":
                    let config = OAuth.Config(authorizeUrl: "", tokenUrl: "", scope: "", extra: [],
                                              clientId: e["clientId"] as? String ?? "",
                                              redirectUri: e["redirectUri"] as? String ?? "")
                    pairs = [
                        ("exchange", .string(OAuth.exchangeForm(
                            config: config, code: e["code"] as? String ?? "",
                            verifier: e["verifier"] as? String ?? ""))),
                        ("kind", .string(kind)),
                        ("name", .string(entryName)),
                        ("refresh", .string(OAuth.refreshForm(
                            config: config, refreshToken: e["refreshToken"] as? String ?? ""))),
                    ]
                case "token_response":
                    let now = e["nowMs"] as? Int ?? 0
                    let state = TokenState.parse(e["prev"] as? String)
                        .applying(responseBody: e["body"] as? String ?? "", nowMs: now)
                    pairs = [
                        ("kind", .string(kind)), ("name", .string(entryName)),
                        ("needsRefresh", .bool(state.needsRefresh(nowMs: now))),
                        ("serialized", .string(state.serialize())),
                        ("state", stateJson(state)),
                        ("usable", .bool(state.isUsable(nowMs: now))),
                    ]
                case "expiry":
                    let now = e["nowMs"] as? Int ?? 0
                    let state = TokenState.parse(e["state"] as? String)
                    pairs = [
                        ("kind", .string(kind)), ("name", .string(entryName)),
                        ("needsRefresh", .bool(state.needsRefresh(nowMs: now))),
                        ("usable", .bool(state.isUsable(nowMs: now))),
                    ]
                case "error":
                    pairs = [
                        ("class", .string(OAuth.classifyTokenError(
                            status: e["status"] as? Int ?? 0,
                            body: e["body"] as? String ?? "").rawValue)),
                        ("kind", .string(kind)), ("name", .string(entryName)),
                    ]
                default:
                    XCTFail("unknown entry type [\(kind)]")
                }
                out.append(.object(pairs))
            }
            let actual = CanonicalJson.render(.array(out))
            if expected != actual {
                failures.append("case [\(name)]\n--- expected ---\n\(expected)\n--- actual ---\n\(actual)")
            }
        }
        if !failures.isEmpty {
            XCTFail("\(failures.count)/\(names.count) oauth cases drifted:\n\n" +
                failures.joined(separator: "\n\n========\n\n"))
        }
    }

    func testGeneratedVerifierIsValidAndDistinct() {
        let a = OAuth.makeCodeVerifier()
        let b = OAuth.makeCodeVerifier()
        XCTAssertEqual(a.count, 64)
        XCTAssertNotEqual(a, b)
        let allowed = Set("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~")
        XCTAssertTrue(a.allSatisfy { allowed.contains($0) })
        XCTAssertEqual(OAuth.makeCodeVerifier(length: 10).count, 43, "下限 43")
        XCTAssertEqual(OAuth.makeCodeVerifier(length: 999).count, 128, "上限 128")
        XCTAssertEqual(OAuth.makeState().count, 43)
    }

    func testRefreshingTokenSource() throws {
        let config = OAuth.Config.gdrive(clientId: "cid", redirectUri: "at.least.crate.ios:/oauth2redirect")
        let fake = FakeTokenServer()
        var persisted: [TokenState] = []
        var sleeps: [Int] = []
        var clock = 1_700_000_000_000
        let source = RefreshingTokenSource(
            config: config, transport: fake,
            state: TokenState(accessToken: "at-1", refreshToken: "rt-1",
                              expiresAtMs: clock + 3_600_000),
            now: { clock }, sleep: { sleeps.append($0) }, persist: { persisted.append($0) })

        XCTAssertEqual(try source.token(), "at-1", "未過期 → 直接用")
        XCTAssertEqual(fake.requests.count, 0, "未過期不打網路")

        clock += 3_600_000 // 過期
        fake.responses = [(200, #"{"access_token":"at-2","expires_in":3599}"#)]
        XCTAssertEqual(try source.token(), "at-2")
        XCTAssertEqual(source.current.refreshToken, "rt-1", "回應無 refresh_token → 沿用")
        XCTAssertEqual(persisted.count, 1)
        let body = String(decoding: fake.requests[0].body, as: UTF8.self)
        XCTAssertEqual(body, "client_id=cid&grant_type=refresh_token&refresh_token=rt-1")
        XCTAssertEqual(fake.requests[0].headers["Content-Type"], "application/x-www-form-urlencoded")

        // 暫時性失敗 → 退避後成功
        clock += 3_600_000
        fake.responses = [(503, ""), (429, ""), (200, #"{"access_token":"at-3","expires_in":3599}"#)]
        XCTAssertEqual(try source.token(), "at-3")
        XCTAssertEqual(sleeps, [1000, 2000], "§2.1 退避序列")

        // invalid_grant → auth（需重新授權），不重試
        clock += 3_600_000
        fake.responses = [(400, #"{"error":"invalid_grant"}"#)]
        XCTAssertThrowsError(try source.token()) { XCTAssertEqual($0 as? ProviderError, .auth) }

        // 無 refresh token → 直接 auth
        let empty = RefreshingTokenSource(config: config, transport: fake, state: .empty,
                                          now: { clock }, sleep: { _ in })
        XCTAssertThrowsError(try empty.token()) { XCTAssertEqual($0 as? ProviderError, .auth) }
    }

    func testExchangeFlow() throws {
        let config = OAuth.Config.dropbox(clientId: "cid", redirectUri: "http://127.0.0.1:7777/callback")
        let fake = FakeTokenServer()
        fake.responses = [(200, #"{"access_token":"at","refresh_token":"rt","expires_in":14400}"#)]
        let source = RefreshingTokenSource(config: config, transport: fake, state: .empty,
                                           now: { 1_700_000_000_000 }, sleep: { _ in })
        let state = try source.exchange(code: "the-code", verifier: "the-verifier")
        XCTAssertEqual(state.accessToken, "at")
        XCTAssertEqual(state.refreshToken, "rt")
        XCTAssertEqual(state.expiresAtMs, 1_700_000_000_000 + 14_400_000)
        XCTAssertTrue(state.isUsable(nowMs: 1_700_000_000_000))
        XCTAssertEqual(String(decoding: fake.requests[0].body, as: UTF8.self),
                       "client_id=cid&code=the-code&code_verifier=the-verifier"
                       + "&grant_type=authorization_code"
                       + "&redirect_uri=http%3A%2F%2F127.0.0.1%3A7777%2Fcallback")
    }

    private func makeConfig(provider: String, clientId: String, redirectUri: String) throws
        -> OAuth.Config {
        switch provider {
        case "gdrive": return .gdrive(clientId: clientId, redirectUri: redirectUri)
        case "dropbox": return .dropbox(clientId: clientId, redirectUri: redirectUri)
        default:
            XCTFail("unknown provider \(provider)")
            throw NSError(domain: "test", code: 1)
        }
    }

    private func stateJson(_ s: TokenState) -> CanonicalJson.JSONValue {
        .object([
            ("accessToken", .string(s.accessToken)),
            ("expiresAtMs", .int(s.expiresAtMs)),
            ("refreshToken", .string(s.refreshToken)),
            ("scope", .string(s.scope)),
        ])
    }

    private func findDir(_ rel: String) -> URL? {
        var dir = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
        while true {
            let c = dir.appendingPathComponent(rel)
            if FileManager.default.fileExists(atPath: c.path) { return c }
            if dir.path == "/" { return nil }
            dir = dir.deletingLastPathComponent()
        }
    }
}

/// 腳本化 token 端點（空佇列 = 200 空 body）。
final class FakeTokenServer: HttpTransport {
    var responses: [(Int, String)] = []
    var requests: [HttpRequest] = []

    func send(_ req: HttpRequest) throws -> HttpResponse {
        requests.append(req)
        guard !responses.isEmpty else { return HttpResponse(status: 200, body: Array("{}".utf8)) }
        let (status, body) = responses.removeFirst()
        return HttpResponse(status: status, body: Array(body.utf8))
    }
}
